### PR TITLE
support constant propagation

### PR DESCRIPTION
**Pull request by XuHuaiyu**
@shenli @hanfei1991 @zimulala @coocood @tiancaiamao 
PTAL
A transformation takes place for expressions like this:

> WHERE column1 = column2 AND column2 = 'x'

the transformed condition becomes:

> WHERE column1 = 'x' AND column2 = 'x'

With constant propagation, we can improve predicate push down
like:
> SELECT * FROM t1 join t2 on t1.a = t2.a where t1.a = 1;

after propagate constant, we get 
> SELECT * FROM t1 join t2 where t2.a = 1 and t1.a = 1;

and we can push down all the conditions.

### CODE REVIEW

File: plan/predicate_push_down.go [hunk 1]
```diff
@@ -45,10 +48,142 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+			expressions := splitDNFItems(conditions[i])
+			newExpression := make([]expression.Expression, 0)
+			for _, v := range expressions {
+				newExpression = append(newExpression, propagateConstant([]expression.Expression{v})...)
+			}
+			conditions[i] = composeDNFConditions(newExpression)
+			isHandled[i] = true
+			continue
+		}
+		if expr.FuncName.L == ast.AndAnd {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: use if ... else if is more readable than if ... continue if ... continue

File: plan/predicate_push_down.go [hunk 2]
```diff
@@ -140,6 +288,7 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
 
 // PredicatePushDown implements LogicalPlan PredicatePushDown interface.
 func (p *Projection) PredicatePushDown(predicates []expression.Expression) (ret []expression.Expression, retPlan LogicalPlan, err error) {
+	//log.Error("Projection")
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: remove this

File: plan/predicate_push_down.go [hunk 3]
```diff
@@ -45,10 +48,142 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+// we propagate constant as the following step:
+// first: "a = 2 and 2 = c and b = 2 and a = 1";
+// next:  "a = 2 and 2 = c and b = 2 and 2 = 1";
+// next:  "0"
+func propagateConstant(conditions []expression.Expression) []expression.Expression {
+	if len(conditions) == 0 {
+		return conditions
+	}
+	isHandled := make([]bool, len(conditions))
+	equalities := make(map[*expression.Column]*expression.Constant, 0)
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Use *Column as map key is not robust. Same column may not have same physical address. Judge it by FromID and PositionID.

File: plan/predicate_push_down.go (hunk 1 above)
Comment (XuHuaiyu) [inline plan/predicate_push_down.go]: Done

File: plan/predicate_push_down.go [hunk 4]
```diff
@@ -47,10 +48,143 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+		if v, ok := equalities[expr.FromID][expr.Position]; ok {
+			return v
+		}
+	case *expression.ScalarFunction:
+		newArgs := make([]expression.Expression, len(expr.Args))
+		for i, arg := range expr.Args {
+			newArgs[i] = constantSubstitute(equalities, arg)
+		}
+		condition, _ = expression.NewFunction(expr.FuncName.L, expr.RetType, newArgs...)
+		return condition
```
Comment (coocood) [inline plan/predicate_push_down.go]: This line can be removed.

File: plan/predicate_push_down.go [hunk 5]
```diff
@@ -47,10 +48,143 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+		}
+		if len(equalities) != 0 {
+			conditions[i] = constantSubstitute(equalities, conditions[i])
+		}
+		expr, ok := conditions[i].(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		// process the included OR conditions recursively to do the same for CNF item.
+		if expr.FuncName.L == ast.OrOr {
```
Comment (coocood) [inline plan/predicate_push_down.go]: We can treat every scalar function the same way except for equal function.

File: expression/expression.go [hunk 6]
```diff
@@ -234,6 +234,16 @@ func (sf *ScalarFunction) MarshalJSON() ([]byte, error) {
 
 // NewFunction creates a new scalar function or constant.
 func NewFunction(funcName string, retType *types.FieldType, args ...Expression) (Expression, error) {
+	if funcName == ast.EQ { // convert a = a to 1
+		arg1, ok1 := args[0].(*Column)
+		arg2, ok2 := args[1].(*Column)
+		if ok1 && ok2 && arg1.Equal(arg2) {
```
Comment (shenli) [inline expression/expression.go]: How about select * from t as t1 join t as t2 on t1.c1 = t2.c?

Comment (shenli) [inline expression/expression.go]: How about select * from t as t1 join t as t2 on t1.c1 = t2.c1?

File: plan/predicate_push_down.go [hunk 7]
```diff
@@ -47,10 +48,143 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+			}
+			isHandled[i] = true
+			i = -1
+		}
+	}
+	// remove useless predicates, such as "1 and a = 1" ==> "a=1".
+	for i := 0; i < len(conditions); i++ {
+		if v, ok := conditions[i].(*expression.Constant); ok {
+			if v.Value.IsNull() {
+				continue
```
Comment (shenli) [inline plan/predicate_push_down.go]: We should break here and change the whole conditions to false.

File: plan/plan_test.go [hunk 8]
```diff
@@ -271,12 +271,16 @@ func (s *testPlanSuite) TestJoinReOrder(c *C) {
 		},
 		{
 			sql:  "select * from t t1, t t2, t t3, t t4, t t5 where t1.a = t5.a and t5.a = t4.a and t4.a = t3.a and t3.a = t2.a and t2.a = t1.a and t1.a = t3.a and t2.a = t4.a and t3.b = 1 and t4.a = 1",
-			best: "LeftHashJoin{LeftHashJoin{LeftHashJoin{LeftHashJoin{Table(t)->Selection->Table(t)}(t3.a,t4.a)->Table(t)}(t4.a,t5.a)->Table(t)}(t5.a,t1.a)(t3.a,t1.a)->Table(t)}(t3.a,t2.a)(t1.a,t2.a)(t4.a,t2.a)->Projection",
+			best: "LeftHashJoin{LeftHashJoin{LeftHashJoin{Table(t)->Selection->Table(t)}->LeftHashJoin{Table(t)->Table(t)}}->Table(t)}->Projection",
```
Comment (shenli) [inline plan/plan_test.go]: Add a test case with null const in condition list.

**Comment by shenli** [comment]
Any update?

**Comment by XuHuaiyu** [comment]
@shenli @hanfei1991 @coocood @zimulala @tiancaiamao PTAL
I add constant propagation support to expression like 
> WHERE a = b and b = c and a > 0 and d = e  and d like 'abc'

the transformed condition becomes:
> WHERE a = b and b = c and d = e and a > 0 and b > 0 and c > 0  and d like 'abc' and e like 'abc'

thus, more conditions can be pushed down.

File: plan/predicate_push_down.go [hunk 9]
```diff
@@ -47,10 +48,281 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+				conditions = []expression.Expression{&expression.Constant{Value: types.NewIntDatum(0)}}
+				return conditions
+			} else {
+				conditions = append(conditions[:i], conditions[i+1:]...)
+				i--
+			}
+		}
+	}
+	if len(conditions) == 0 {
+		conditions = []expression.Expression{&expression.Constant{Value: types.NewIntDatum(1)}}
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Why do this ?

File: plan/predicate_push_down.go [hunk 10]
```diff
@@ -47,10 +48,281 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+			position = col.Position
+			if _, ok = equalities[fromID]; !ok {
+				equalities[fromID] = make(map[int]*expression.Constant, 1)
+			}
+			equalities[fromID][position] = val
+			isHandled[i] = true
+			i = -1
+		}
+	}
+	for i := 0; i < len(conditions); i++ { // remove useless predicates, such as "1 and a = 1" ==> "a=1".
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: It's no need to do this. If the condition is a forever-false expression, we will change table scan to dummy table scan in physical plan transforming phase.

File: plan/predicate_push_down.go (hunk 9 above)
Comment (XuHuaiyu) [inline plan/predicate_push_down.go]: For expression, like expression  "2 = 2 and 1 = 1", I remove all of them in previous code, but now that these code is proved to be useless, there is no need to do this, too.

File: plan/predicate_push_down.go [hunk 11]
```diff
@@ -47,10 +48,281 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+	// next:  "1 = b and 1 = c and 1 = 1 and a = 1";
+	// next:  "1 = b and 1 = c and a = 1";
+
+	// e.g for condition: "a = b and b = c and b = 2 and a = 1";
+	// we propagate constant as the following step:
+	// first: "a = 2 and 2 = c and b = 2 and a = 1";
+	// next:  "a = 2 and 2 = c and b = 2 and 2 = 1";
+	// next:  "0"
+	isHandled := make([]bool, len(conditions))
+	equalities := make(map[string]map[int]*expression.Constant, 0) // transitive equality predicates between one column and one constant
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: It's ugly. I think it's better to wrap it to a data structrue.

File: plan/predicate_push_down.go [hunk 12]
```diff
@@ -47,10 +48,281 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+		}
+		if len(equalities) != 0 {
+			conditions[i] = constantSubstitute(equalities, conditions[i])
+		}
+		expr, ok := conditions[i].(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		// process the included OR conditions recursively to do the same for CNF item.
+		if expr.FuncName.L == ast.OrOr {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: switch expr.FuncName.L is better.

File: plan/predicate_push_down.go [hunk 13]
```diff
@@ -47,10 +48,281 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+			expr.Args[i] = constantSubstitute(equalities, arg)
+		}
+		condition, _ = expression.NewFunction(expr.FuncName.L, expr.RetType, expr.Args...)
+		return condition
+	}
+	return condition
+}
+
+// composeDNFCondition composes DNF items into a balanced deep DNF tree.
+func composeDNFConditions(conditions []expression.Expression) expression.Expression {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: You can move this to expression package.

File: plan/predicate_push_down.go (hunk 11 above)
Comment (XuHuaiyu) [inline plan/predicate_push_down.go]: You mean this? For more readable.
``` go
type transitiveEqualityPredicate map[string]map[int]*expression.Constant  
equalities := make(transitiveEqualityPredicate, 0)
```

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

File: plan/predicate_push_down.go [hunk 14]
```diff
@@ -47,10 +48,248 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+			col      *expression.Column
+			equalCol *expression.Column // the root column corresponding to a column in a multiple equality predicate.
+			fromID   string
+			position int
+			val      *expression.Constant
+			isColumn bool
+			funcName string
+		)
+		left, ok1 := expr.Args[0].(*expression.Constant)
+		right, ok2 := expr.Args[1].(*expression.Constant)
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: What if the function only has one arg?

File: plan/predicate_push_down.go [hunk 15]
```diff
@@ -47,10 +48,248 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+		if !ok {
+			continue
+		}
+		equalCol, ok = multipleEqualities[col]
+		if !ok { // no need to propagate inequality predicates whose column is only equal to itself.
+			continue
+		}
+		fromID, position = equalCol.FromID, equalCol.Position
+		if _, ok = inequalities[fromID]; !ok {
+			inequalities[fromID] = make(map[int]map[string][]*expression.Constant, 1)
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: It's better to encode (FromID, PositionID) as a HashCode binary key by using EncodeValue

File: plan/predicate_push_down.go [hunk 16]
```diff
@@ -47,10 +48,248 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+			expr.Args[i] = constantSubstitute(equalities, arg)
+		}
+		condition, _ = expression.NewFunction(expr.FuncName.L, expr.RetType, expr.Args...)
+		return condition
+	}
+	return condition
+}
+
+// DNF means disjunctive normal form, e.g. a or b or c.
+func splitDNFItems(expr expression.Expression) []expression.Expression {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: It's better to move this to expression package.

**Comment by XuHuaiyu** [comment]
PTAL
@hanfei1991

File: executor/builder.go [hunk 17]
```diff
@@ -447,9 +447,9 @@ func (b *executorBuilder) buildSemiJoin(v *plan.PhysicalHashSemiJoin) Executor {
 	}
 	e := &HashSemiJoinExec{
 		schema:       v.GetSchema(),
-		otherFilter:  expression.ComposeCNFCondition(v.OtherConditions),
-		bigFilter:    expression.ComposeCNFCondition(v.LeftConditions),
-		smallFilter:  expression.ComposeCNFCondition(v.RightConditions),
+		otherFilter:  expression.ComposeConditionWithBinaryOp(v.OtherConditions, ast.AndAnd),
```
Comment (hanfei1991) [inline executor/builder.go]: I think ComposeCNF / DNF is better, you can call composeConditionWithBinaryOp inside expression package.

File: plan/predicate_push_down.go [hunk 18]
```diff
@@ -47,10 +49,229 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+			left, ok1 := expr.Args[0].(*expression.Constant)
+			right, ok2 := expr.Args[1].(*expression.Constant)
+			if col, isColumn = expr.Args[0].(*expression.Column); ok2 && isColumn {
+				val = right
+			} else if col, isColumn = expr.Args[1].(*expression.Column); ok1 && isColumn {
+				val = left
+			} else {
+				continue
+			}
+			bytes, _ = codec.EncodeValue(bytes, types.NewStringDatum(col.FromID), types.NewIntDatum(int64(col.Position)))
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: It's better to implement a HashCode function for Expression interface

File: plan/predicate_push_down.go [hunk 19]
```diff
@@ -47,10 +49,229 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+			equalities[string(bytes)] = val
+			isHandled[i] = true
+			i = -1
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Why set i as -1?

Comment (XuHuaiyu) [inline plan/predicate_push_down.go]: e.g.  "a = b and a = 1 and b = c"
when find "a = 1", set all "a" to "1" in the condition except "a=1"
so here set i = -1 to restart the loop to propagate equality predicates

File: plan/predicate_push_down.go [hunk 20]
```diff
@@ -47,10 +49,229 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+	}
+	inequalityFuncs := map[string]string{
+		ast.LT:   ast.LT,
+		ast.GT:   ast.GT,
+		ast.LE:   ast.LE,
+		ast.GE:   ast.GE,
+		ast.NE:   ast.NE,
+		ast.Like: ast.Like,
+	}
+	type transitiveInEqualityPredicate map[string]map[string][]*expression.Constant // transitive inequality predicates between one column and one constant.
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: change map[string][]*expressionConstant as a List like []*struct. It's ok to use a structure without dedup function.

File: plan/predicate_push_down.go (hunk 19 above)
Comment (hanfei1991) [inline plan/predicate_push_down.go]: I think it is more clear to split it to two loops. one loop find equal conditions, another one propagates constants. And outside them there is a loop to drive them util no more new equal conds appears.

**Comment by shenli** [comment]
@XuHuaiyu CI failed

File: expression/expression.go [hunk 21]
```diff
@@ -40,6 +41,9 @@ type Expression interface {
 
 	// DeepCopy copies an expression totally.
 	DeepCopy() Expression
+
+	// HashCode create the hashcode for expression
+	HashCode() string
```
Comment (hanfei1991) [inline expression/expression.go]: return []byte is more efficient.

File: expression/expression.go [hunk 22]
```diff
@@ -361,25 +383,57 @@ func (c *Constant) Eval(_ []types.Datum, _ context.Context) (types.Datum, error)
...
+
 // Assignment represents a set assignment in Update, such as
 // Update t set c1 = hex(12), c2 = c3 where c2 = 1
 type Assignment struct {
 	Col  *Column
 	Expr Expression
 }
+
+// splitNormalFormItems split CNF(conjunctive normal form) like "a and b and c", or DNF(disjunctive normal form) like "a or b or c"
+func SplitNormalFormItems(onExpr Expression, funcName string) []Expression {
```
Comment (hanfei1991) [inline expression/expression.go]: Implement splitCNF/splitDNF like composeCNF/DNF

File: plan/predicate_push_down.go [hunk 23]
```diff
@@ -47,10 +49,225 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+		ast.LE:   ast.LE,
+		ast.GE:   ast.GE,
+		ast.NE:   ast.NE,
+		ast.Like: ast.Like,
+	}
+	type inequalityFactor struct {
+		FuncName string
+		Factor   []*expression.Constant
+	}
+	type transitiveInEqualityPredicate map[string][]inequalityFactor // transitive inequality predicates between one column and one constant.
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: why not *inequalityFactor?

Comment (XuHuaiyu) [inline plan/predicate_push_down.go]: Why here need *inequalityFactor,
it would not be big slice, and value of the member attributes would not be modified.

**Comment by XuHuaiyu** [comment]
@hanfei1991 PTAL

**Comment by shenli** [comment]
PTAL @coocood @zimulala

File: plan/predicate_push_down.go [hunk 24]
```diff
@@ -47,10 +48,224 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+	// next:  "1 = b and 1 = c and a = 1";
+
+	// e.g for condition: "a = b and b = c and b = 2 and a = 1";
+	// we propagate constant as the following step:
+	// first: "a = 2 and 2 = c and b = 2 and a = 1";
+	// next:  "a = 2 and 2 = c and b = 2 and 2 = 1";
+	// next:  "0"
+	isHandled := make([]bool, len(conditions))
+	type transitiveEqualityPredicate map[string]*expression.Constant // transitive equality predicates between one column and one constant
+	for true {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: for true { is equal to for {

File: plan/predicate_push_down.go [hunk 25]
```diff
@@ -47,10 +48,224 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+			for k, v := range multipleEqualities {
+				if v.Equal(rootOfRightExpr) {
+					multipleEqualities[k] = v
+				}
+			}
+		}
+	}
+	for _, cond := range conditions { // build multiple equality predicates.
+		expr, ok := cond.(*expression.ScalarFunction)
+		if !ok || expr.FuncName.L != ast.EQ {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: using ok && expr.FuncName.L == ast.EQ is better.

File: plan/predicate_push_down.go [hunk 26]
```diff
@@ -47,10 +48,224 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+	// 1. build multiple equality predicates(mep):
+	//    =(a, b, c, d), =(g, h).
+	// 2. extract inequality predicates between one constant and one column,
+	//    and rewrite them using the root column of a multiple equality predicate:
+	//    b > 0, e != 0, g like 'abc' ==> a > 0, g like 'abc'.
+	//    ATTENTION: here column 'e' doesn't belong to any mep, so we skip "e != 0".
+	// 3. propagate constants in these inequality predicates, and we finally get:
+	//    "a = b and c = d and a = c and e = f and g = h and e != 0 and a > 0 and b > 0 and c > 0 and d > 0 and g like 'abc' and h like 'abc' ".
+	multipleEqualities := make(map[*expression.Column]*expression.Column, 0)
+	union := func(leftExpr *expression.Column, rightExpr *expression.Column) {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: make it a public function.

File: expression/expression.go [hunk 27]
```diff
@@ -361,25 +385,67 @@ func (c *Constant) Eval(_ []types.Datum, _ context.Context) (types.Datum, error)
...
 	if length == 1 {
 		return conditions[0]
 	}
-	expr, _ := NewFunction(ast.AndAnd,
+	expr, _ := NewFunction(funcName,
 		types.NewFieldType(mysql.TypeTiny),
-		ComposeCNFCondition(conditions[length/2:]),
-		ComposeCNFCondition(conditions[:length/2]))
+		composeConditionWithBinaryOp(conditions[length/2:], funcName),
+		composeConditionWithBinaryOp(conditions[:length/2], funcName))
```
Comment (hanfei1991) [inline expression/expression.go]: Why not use : length/2 firstly?

**Comment by hanfei1991** [comment]
Write special unit test for constant propagation by comparing expression string, but not only add test for ppd.

File: expression/expression.go [hunk 28]
```diff
@@ -328,6 +338,20 @@ func (sf *ScalarFunction) Eval(row []types.Datum, ctx context.Context) (types.Da
...
+// HashCode implements Expression interface.
+func (sf *ScalarFunction) HashCode() []byte {
+	var bytes []byte
+	v := make([]types.Datum, 0, len(sf.Args)+1)
+	bytes, _ = codec.EncodeValue(bytes, types.NewStringDatum(sf.FuncName.L))
+	v = append(v, types.NewBytesDatum(bytes))
+	for _, arg := range sf.Args {
+		v = append(v, types.NewBytesDatum(arg.HashCode()))
+	}
+	bytes = make([]byte, 0)
```
Comment (overvenus) [inline expression/expression.go]: How about `bytes[:0]` ?

File: expression/expression.go (hunk 27 above)
Comment (XuHuaiyu) [inline expression/expression.go]: Why? Is there any different?

Comment (XuHuaiyu) [inline expression/expression.go]: Ok, I'll modify it.
For CNF and DNF, here is no difference.

File: expression/expression.go (hunk 28 above)
Comment (XuHuaiyu) [inline expression/expression.go]: Yeah, it would be better.

File: plan/plan_test.go [hunk 29]
```diff
@@ -1054,3 +1056,67 @@ func check(p Plan, c *C, ans map[string][]string, comment CommentInterface) {
...
 }
+
+func (s *testPlanSuite) TestConstantPropagation(c *C) {
+	defer testleak.AfterTest(c)()
+	cases := []struct {
+		sql   string
+		after string
+	}{
+		{
+			sql:   "select * from t where a = b and b = c and c = d and d = 1",
```
Comment (hanfei1991) [inline plan/plan_test.go]: Why add select * for each test in map? you can add it in the for loop.

**Comment by hanfei1991** [comment]
LGTM @shenli @coocood @zimulala PTAL

File: expression/expression.go [hunk 30]
```diff
@@ -83,7 +87,7 @@ type Column struct {
 // Equal checks if two columns are equal
 func (col *Column) Equal(expr Expression) bool {
 	if newCol, ok := expr.(*Column); ok {
-		return col.FromID == newCol.FromID && col.ColName == newCol.ColName
+		return string(col.HashCode()) == string(newCol.HashCode())
```
Comment (coocood) [inline expression/expression.go]: Why not use `bytes.Equal`?

File: plan/plan_test.go [hunk 31]
```diff
@@ -130,6 +132,16 @@ func (s *testPlanSuite) TestPredicatePushDown(c *C) {
 			best:  "Join{DataScan(t)->DataScan(t)}->Projection",
 		},
 		{
+			sql:   "select * from t where a = b and b = 1 and a = null and c = d and c > 2 and c != 4",
+			first: "DataScan(t)->Selection->Projection",
+			best:  "DataScan(t)->Selection->Projection",
```
Comment (coocood) [inline plan/plan_test.go]: This case can be removed.

File: plan/predicate_push_down.go [hunk 32]
```diff
@@ -47,10 +48,225 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
 	return expr
 }
 
+// propagateConstant propagate constant values of equality predicates and inequality predicates in a condition.
+func propagateConstant(conditions []expression.Expression) []expression.Expression {
```
Comment (coocood) [inline plan/predicate_push_down.go]: define arguments as `conditions ...expression.Expression`, so you can pass in one expression.

Comment (coocood) [inline plan/predicate_push_down.go]: Never mind, slice argument are used more frequently.

File: plan/predicate_push_down.go [hunk 33]
```diff
@@ -47,10 +48,225 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+				newExpression := propagateConstant(expression.SplitCNFItems(conditions[i]))
+				conditions[i] = expression.ComposeCNFCondition(newExpression)
+				isHandled[i] = true
+			case ast.EQ:
+				var (
+					col      *expression.Column
+					val      *expression.Constant
+					isColumn bool
+				)
+				left, ok1 := expr.Args[0].(*expression.Constant)
```
Comment (coocood) [inline plan/predicate_push_down.go]: rename `left`, `ok1`, `right`, `ok2` to `leftConst`, `leftIsConst`, `rightConst`, `rightIsConst`
and add lines
```
leftCol, leftIsCol := expr.Args[0].(*expression.Column)
rightCol, rightIsCol := expr.Args[1].(*expression.Column)
```
could be more readable.

File: plan/predicate_push_down.go [hunk 34]
```diff
@@ -47,10 +48,225 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+	// next:  "1 = b and 1 = c and c = 1 and a = 1";
+	// next:  "1 = b and 1 = c and 1 = 1 and a = 1";
+	// next:  "1 = b and 1 = c and a = 1";
+
+	// e.g for condition: "a = b and b = c and b = 2 and a = 1";
+	// we propagate constant as the following step:
+	// first: "a = 2 and 2 = c and b = 2 and a = 1";
+	// next:  "a = 2 and 2 = c and b = 2 and 2 = 1";
+	// next:  "0"
+	isHandled := make([]bool, len(conditions))
```
Comment (coocood) [inline plan/predicate_push_down.go]: Rename `isHandle` to `isSource` can be more readable, as every propagation has a source column equals constant expression, which is used to propagate that column appeared in other expressions.

File: plan/predicate_push_down.go [hunk 35]
```diff
@@ -47,10 +48,225 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+		if !ok {
+			continue
+		}
+		funcName, ok = inequalityFuncs[expr.FuncName.L]
+		if !ok {
+			continue
+		}
+		left, ok1 := expr.Args[0].(*expression.Constant)
+		right, ok2 := expr.Args[1].(*expression.Constant)
+		if column, isColumn = expr.Args[0].(*expression.Column); ok2 && isColumn {
```
Comment (coocood) [inline plan/predicate_push_down.go]: Save as above comment.

File: plan/predicate_push_down.go [hunk 36]
```diff
@@ -47,10 +48,225 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
...
+		} else {
+			continue
+		}
+		equalCol, ok = multipleEqualities[column]
+		if !ok { // no need to propagate inequality predicates whose column is only equal to itself.
+			continue
+		}
+		colHashCode := string(equalCol.HashCode())
+		if _, ok = inequalities[colHashCode]; !ok {
+			inequalities[colHashCode] = make([]inequalityFactor, 0)
```
Comment (coocood) [inline plan/predicate_push_down.go]: We don't need to make an empty slice before append.

**Comment by coocood** [comment]
Can we do `outerJoinSimplify` first, then `propagateConstant`, then `predicatePushDown`?
@hanfei1991

**Comment by hanfei1991** [comment]
@coocood It's better to do them together to do optimizer as deeply as possible.

**Comment by coocood** [comment]
@hanfei1991  after propagateConstant, join reorder result can be more optimized.

**Comment by hanfei1991** [comment]
@coocood Currently, our join order framework only support a,b,c, where a.id = b.id ... It's very simple.
In future we will extract all the conditions around every joins and do reorder after we considering cost info.

**Comment by coocood** [comment]
LGTM

**Pull request merged**